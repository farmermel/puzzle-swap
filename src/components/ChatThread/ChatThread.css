.chat-thread {
  display: flex;
  justify-content: center;
  padding: 170px 0 0 0;
  height: 100vh;
  overflow-y: scroll;
}

.chat-form {
  position: fixed;
  bottom: 8px;
  padding: 0;
  align-items: center;
  display: flex;
  flex-direction: row;
  height: 36px;
  width: 100%;
}

.chat-form input {
  margin: 0;
}

.submit-chat {
  width: 100px;
}

.chat-members {
  position: fixed;
  top: 135px;
  left: 40px;
}