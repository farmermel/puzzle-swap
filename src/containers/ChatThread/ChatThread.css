.chat-thread {
  display: flex;
  justify-content: center;
  padding: 170px 0 0 0;
  height: 100vh;
  overflow-y: scroll;
}

.chat-form {
  position: fixed;
  bottom: 8px;
  padding: 0;
  align-items: center;
  display: flex;
  flex-direction: row;
  height: 36px;
  width: 100%;
}

.chat-form input {
  margin: 0;
  height: 45px;
}

.submit-chat {
  width: 100px;
}

.chat-members {
  position: fixed;
  top: 135px;
  left: 40px;
}

.speech-right {
  position: relative;
  background: #fff;
  border-radius: .4em;
  padding: 5px 27px;
  margin: 20px 0;
}

.speech-right:after {
  content: '';
  position: absolute;
  right: 0;
  top: 50%;
  width: 0;
  height: 0;
  border: 36px solid transparent;
  border-left-color: #fff;
  border-right: 0;
  border-bottom: 0;
  margin-top: -18px;
  margin-right: -36px;
}

.speech-left {
  position: relative;
  background: #dedede;
  border-radius: .4em;
  padding: 5px 27px;
  box-shadow: 0 1px 2px #999;

  margin: 20px 0;
}

.speech-left:after {
  content: '';
  position: absolute;
  left: 0;
  top: 50%;
  width: 0;
  height: 0;
  border: 36px solid transparent;
  border-right-color: #dedede;
  box-shadow: 0 3px 2px -2px #999;
  /*box-shadow: 0 1px 2px #999;*/
  border-left: 0;
  border-bottom: 0;
  margin-top: -18px;
  margin-left: -36px;
}