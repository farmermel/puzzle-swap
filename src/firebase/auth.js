import { auth } from './firebase';

export const doCreateUserWithEmailAndPassword = (email, password) =>
  auth.createUserWithEmailAndPassword(email, password);

export const doSignInWithEmailAndPassword = (email, password) => 
  auth.signInWithEmailAndPassword(email, password);

export const onAuthStateChanged = user => 
  auth.onAuthStateChanged(user);

export const doSignOut = () => 
  auth.signOut();