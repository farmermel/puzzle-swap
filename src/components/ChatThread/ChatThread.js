import React, { Component } from 'react';
import { connect } from 'react-redux';
import { db } from '../../firebase';
import './ChatThread.css';

class ChatThread extends Component {
  constructor() {
    super();
    this.state = {
      message: '',
      messagesToRender: []
    }
  }

  async componentDidMount() {
    const { chat } = this.props;
    const messagesSnap = await db.watchData(`messages/${chat.chatId}`)
    messagesSnap.on('value', snapshot => {
      this.renderMessages(snapshot.val())
    })
  }

  getMembers = members => {
    const membersArr = Object.keys(members).map( member => (
      members[member]
    ))
    return membersArr.join(', ')
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const { user, chat } = this.props;
      const date = Date()
      const postDB = {
        username: user.username,
        uid: user.uid,
        message: this.state.message,
        timeStamp: date.toString()
      }
      const firebaseKey = db.getFirebaseKey(`messages/${chat.chatId}`);
      let updates = {};
      updates[`messages/${chat.chatId}/${firebaseKey}`] = postDB;
      updates[`chats/${chat.chatId}/lastMessage`] = this.state.message;
      updates[`chats/${chat.chatId}/timeStamp`] = postDB.timeStamp;
      this.setState({ message: '' })
      await db.postUpdate(updates);
    } catch (error) {
      console.log(error)
    }
  }

  // getMessages = async () => {
  //   // const { chat } = this.props;
  //   // const messagesSnap = await db.watchData(`messages/${chat.chatId}`)
  //   // messagesSnap.on('value', snapshot => {
  //   //   this.renderMessages(snapshot.val())
  //   // })
  // }

  renderMessages = messages => {
    const messagesToRender = messages
      ? Object.keys(messages).map( mkey => {
          return (
            <article>
              <div>
                <p>{messages[mkey].timeStamp}</p>
                <h3>{messages[mkey].username}</h3>
              </div>
              <p>{messages[mkey].message}</p>
            </article>
          )
        })
      : <h3>Introduce yourself!</h3>
    this.setState({ messagesToRender })
  }

  render() {
    const { chat } = this.props;
    return (
      <section className='chat-thread'>
        <h3 className='chat-members'>Members: <span>{this.getMembers(chat.members)}</span></h3>
        <section className='messages-wrap'>
          {this.state.messagesToRender}
        </section>
        <form onSubmit={ this.handleSubmit }
              className='chat-form'>
          <input type='text'
                 name='message'
                 value={ this.state.message }
                 onChange={ this.handleChange } />
          <button className='submit-chat'>Send</button>
        </form>
      </section>
    ) 
  }
}

const mapStateToProps = state => ({
  user: state.user
})

export default connect(mapStateToProps, null)(ChatThread);