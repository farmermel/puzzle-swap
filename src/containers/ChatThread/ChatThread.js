import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { db } from '../../firebase';
import { hasErrored } from '../../actions/hasErrored';
import PropTypes from 'prop-types';
import moment from 'moment';
import './ChatThread.css';

export class ChatThread extends Component {
  constructor() {
    super();
    this.state = {
      message: '',
      messagesToRender: []
    }
  }

  async componentDidMount() {
    const { chat, hasErrored } = this.props;
    try {
      const messagesSnap = await db.watchData(`messages/${chat.chatId}`);
      messagesSnap.on('value', async snapshot => {
        await this.renderMessages(snapshot.val())
      })
    } catch (error) {
      hasErrored(error.message);
    }
  }

  getRecipientName = members => {
    const { user } = this.props;
    const recipientId = Object.keys(members).find( member => {
      return member !== user.uid
    });
    return members[recipientId].username;
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  formatTime = () => {
    return moment().format('h:mma, MMMM Do');
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { user, chat, hasErrored } = this.props;
    try {
      const postDB = {
        username: user.username,
        uid: user.uid,
        message: this.state.message,
        timeStamp: this.formatTime()
      }
      const firebaseKey = db.getFirebaseKey(`messages/${chat.chatId}`);
      let updates = {};
      updates[`messages/${chat.chatId}/${firebaseKey}`] = postDB;
      updates[`chats/${chat.chatId}/lastMessage`] = this.state.message;
      updates[`chats/${chat.chatId}/timeStamp`] = postDB.timeStamp;
      this.setState({ message: '' })
      await db.postUpdate(updates);
    } catch (error) {
      hasErrored(error.message);
    }
  }

  renderMessages = messages => {
    const colorObj = this.determineMessageColor();
    const messagesToRender = messages
      ? Object.keys(messages).map( mkey => {
          return (
            <article className={colorObj[messages[mkey].uid]}
                     key={mkey}>
              <div className='chatthread-details'>
                <p>{messages[mkey].timeStamp}</p>
                <p>{messages[mkey].username}</p>
              </div>
              <p>{messages[mkey].message}</p>
            </article>
          )
        })
      : <h3 className='no-messages'>Introduce yourself!</h3>
    this.setState({ messagesToRender })
  }

  determineMessageColor = () => {
    const { chat, user } = this.props;
    const members = Object.keys(chat.members);
    const recipient = members.find( member => (
      user.uid !== member
    ));
    return {
      [user.uid]: 'speech-right',
      [recipient]: 'speech-left'
    }
  }

  render() {
    this.formatTime();
    const { chat } = this.props;
    return (
      <section className='chat-thread'>
        <h3 className='chat-members'>Your messages with <span>{this.getRecipientName(chat.members)}</span></h3>
        <section className='messages-wrap'>
          {this.state.messagesToRender}
        </section>
        <form onSubmit={ this.handleSubmit }
              className='chat-form'>
          <input type='text'
                 name='message'
                 value={ this.state.message }
                 onChange={ this.handleChange } />
          <button type='submit'
                  value='Send'
                  className='submit-chat'>Send</button>
        </form>
      </section>
    ) 
  }
}

export const mapStateToProps = state => ({
  user: state.user
})

export const mapDispatchToProps = dispatch => ({
  hasErrored: message => dispatch(hasErrored(message))
})

ChatThread.propTypes = {
  user: PropTypes.shape({
    uid: PropTypes.number,
    username: PropTypes.string
  }),
  chat: PropTypes.shape({
    members: PropTypes.objectOf(PropTypes.objectOf(PropTypes.string)),
    chatId: PropTypes.string
  }),
  hasErrored: PropTypes.func.isRequired
}

export default withRouter(connect(mapStateToProps, null)(ChatThread));