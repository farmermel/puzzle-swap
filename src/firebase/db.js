import { db } from './firebase';

export const doCreateUser = (id, username, email) => 
  db.ref(`users/${id}`).set({
    username,
    email
  });

// export const onceGetUsers = () => 
//   db.ref('users').once('value');

export const getOnce = path =>
  db.ref(path).once('value');

export const getFirebaseKey = childNode => 
  db.ref().child(childNode).push().key;

export const postUpdate = updates => 
  db.ref().update(updates);

export const watchData = watchNode => 
  db.ref(watchNode);

// export const getData = watchNode => 
//   watchData(watchNode).on('value', (snapshot) => {
//     return snapshot.val()
//   })