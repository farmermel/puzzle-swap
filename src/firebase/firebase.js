import * as firebase from 'firebase';
import { firebaseKey,
  fbAuthDomain,
  fbdbUrl,
  fbProjId,
  fbMessagingSenderId } from '../helpers/apiKeys';

const config = {
  apiKey: firebaseKey,
  authDomain: fbAuthDomain,
  databaseURL: fbdbUrl,
  projectId: fbProjId,
  storageBucket: "",
  messagingSenderId: fbMessagingSenderId
};

if (!firebase.apps.length) {
  firebase.initializeApp(config);
}

const auth = firebase.auth();

const db = firebase.database();

export {
  db,
  auth,
};