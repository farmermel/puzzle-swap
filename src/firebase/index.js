import * as auth from './auth';
import * as db from './db';
import * as firebase from './firebase';
import * as store from './store';

export {
  auth,
  db,
  store,
  firebase,
};