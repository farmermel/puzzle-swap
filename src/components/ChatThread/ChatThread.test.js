import React from 'react';
import { shallow } from 'enzyme';
import { ChatThread, mapStateToProps } from './ChatThread';
import { db } from '../../firebase';

describe('ChatThread', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ChatThread user={{username: 'Nymeria', uid: 4}}
                                  chat={{members: {1: 'Nymeria', 2: 'Arya'}, chatId: '4'}}/>,
                      {disableLifecycleMethods: true})
  })
  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('has default state', () => {
    const expected = {
      message: '',
      messagesToRender: [],
      error: null
    };
    expect(wrapper.instance().state).toEqual(expected);
  })

  describe('componentDidMount', () => {
    beforeEach(() => {
      db.watchData = jest.fn().mockImplementation(() => {
        return {
          on: jest.fn()
        }
      })
      wrapper.instance().renderMessages = jest.fn();
    })

    afterAll(() => {
      db.getFirebaseKey.mockClear();
    })

    it('calls watchData method on db with messages/chatid', () => {
      expect(db.watchData).not.toHaveBeenCalled();
      wrapper.instance().componentDidMount();
      expect(db.watchData).toHaveBeenCalledWith('messages/4');
    })
  })

  describe('getMembers', () => {
    it('returns a comma seperated string of the values from an object passed in', () => {
      const mockMembers = {
        4: 'Nymeria',
        5: 'Arya'
      };
      const expected = 'Nymeria, Arya';
    })
  })

  describe('handleChange', () => {
    it('sets state with name and value of event passed in', () => {
      const mockEvent = {
        target: {
          name: 'message',
          value: 'Valhar morgulis'
        }
      }
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.instance().state.message).toEqual('Valhar morgulis');
    })
  })

  describe('handleSubmit', () => {
    let mockEvent;
    beforeEach(() => {
      db.getFirebaseKey = jest.fn().mockImplementation(() => {
        return 4
      })
      db.postUpdate = jest.fn();
      mockEvent = {
        preventDefault: jest.fn()
      }
    })

    it('prevents default of form submission', () => {
      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
      wrapper.instance().handleSubmit(mockEvent);
      expect(mockEvent.preventDefault).toHaveBeenCalled();
    })

    it('calls getFirebaseKey method on db', () => {
      expect(db.getFirebaseKey).not.toHaveBeenCalled();
      wrapper.instance().handleSubmit(mockEvent);
      expect(db.getFirebaseKey).toHaveBeenCalled();
    })

    it('clears message in state', () => {
      wrapper.instance().setState({ message: 'valhar morgulis' });
      expect(wrapper.instance().state.message).toEqual('valhar morgulis');
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().state.message).toEqual('');
    })

    it('calls postUpdate on database with an updates object containing three keys for different updates', () => {
      expect(db.postUpdate).not.toHaveBeenCalled();
      wrapper.instance().handleSubmit(mockEvent);
      expect(db.postUpdate).toHaveBeenCalled();
    })

    it('sets state with error if something goes wrong', () => {
      db.postUpdate = jest.fn().mockImplementation(() => {
        throw new Error('valhar doharis')
      })
      wrapper.instance().handleSubmit(mockEvent);
      expect(wrapper.instance().state.error).toEqual('valhar doharis');
    })
  })

  describe('renderMessages', () => {
    let mockMessages;
    beforeAll(() => {
      wrapper.instance().determineMessageColor = jest.fn().mockImplementation(() => {
        return { 
          5: 'yellow',
          6: 'green'
        }
      })
      mockMessages = {
        5: {
          timeStamp: 4,
          username: 'Nymeria',
          message: 'Arroooooo',
          uid: 5
        },
        6: {
          timeStamp: 7,
          username: 'Aria',
          message: 'Needle!',
          uid: 6
        }
      }
    })

    it('sets state with messages to render if there are messages', () => {
      const expected = [<article className={undefined}><div><p>4</p><h3>Nymeria</h3></div><p>Arroooooo</p></article>, <article className={undefined}><div><p>7</p><h3>Aria</h3></div><p>Needle!</p></article>]
      wrapper.instance().renderMessages(mockMessages);
      expect(typeof wrapper.instance().state.messagesToRender).toEqual(typeof expected);
    })

    it('sets state with instructions to introduce yourself if there are no messages', () => {
      const expected = <h3>Introduce yourself!</h3>
      wrapper.instance().renderMessages(undefined);
      expect(wrapper.instance().state.messagesToRender).toEqual(expected);
    })
  })

  describe('determineMessageColor', () => {
    it('returns an object with member ids as keys and classnames as values', () => {
      const expected = {1: 'speech-right', 2: 'speech-left'};
      expect(wrapper.instance().determineMessageColor()).toEqual(expected)
    })
  })

  describe('mapStateToProps', () => {
    it('maps state to props', () => {
      const mockState = {
        user: 'Nymeria'
      }
      const mapped = mapStateToProps(mockState);
      expect(mapped.user).toEqual(mockState.user);
    })
  })
})