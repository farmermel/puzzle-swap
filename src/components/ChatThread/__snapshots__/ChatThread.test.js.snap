// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatThread matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ChatThread
    chat={
        Object {
            "chatId": "4",
            "members": Object {
              "1": "Nymeria",
              "2": "Arya",
            },
          }
    }
    user={
        Object {
            "uid": 4,
            "username": "Nymeria",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="chat-members"
>
          Members: 
          <span>
                    Nymeria, Arya
          </span>
</h3>,
        <section
          className="messages-wrap"
>
          
</section>,
        <form
          className="chat-form"
          onSubmit={[Function]}
>
          <input
                    name="message"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <button
                    className="submit-chat"
          >
                    Send
          </button>
</form>,
      ],
      "className": "chat-thread",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Members: ",
            <span>
              Nymeria, Arya
</span>,
          ],
          "className": "chat-members",
        },
        "ref": null,
        "rendered": Array [
          "Members: ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Nymeria, Arya",
            },
            "ref": null,
            "rendered": "Nymeria, Arya",
            "type": "span",
          },
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "messages-wrap",
        },
        "ref": null,
        "rendered": Array [],
        "type": "section",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="message"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <button
              className="submit-chat"
>
              Send
</button>,
          ],
          "className": "chat-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "message",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Send",
              "className": "submit-chat",
            },
            "ref": null,
            "rendered": "Send",
            "type": "button",
          },
        ],
        "type": "form",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="chat-members"
>
            Members: 
            <span>
                        Nymeria, Arya
            </span>
</h3>,
          <section
            className="messages-wrap"
>
            
</section>,
          <form
            className="chat-form"
            onSubmit={[Function]}
>
            <input
                        name="message"
                        onChange={[Function]}
                        type="text"
                        value=""
            />
            <button
                        className="submit-chat"
            >
                        Send
            </button>
</form>,
        ],
        "className": "chat-thread",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Members: ",
              <span>
                Nymeria, Arya
</span>,
            ],
            "className": "chat-members",
          },
          "ref": null,
          "rendered": Array [
            "Members: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Nymeria, Arya",
              },
              "ref": null,
              "rendered": "Nymeria, Arya",
              "type": "span",
            },
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "messages-wrap",
          },
          "ref": null,
          "rendered": Array [],
          "type": "section",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="message"
                onChange={[Function]}
                type="text"
                value=""
/>,
              <button
                className="submit-chat"
>
                Send
</button>,
            ],
            "className": "chat-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "message",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Send",
                "className": "submit-chat",
              },
              "ref": null,
              "rendered": "Send",
              "type": "button",
            },
          ],
          "type": "form",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
